File: server/internal/model/agent.go
package model

import "time"

// AgentStatus 用类型常量集中定义，避免拼写错误、便于 IDE 补全。
// 底层是 string，DB scan / JSON 序列化均按 string 字面值处理。
//
// 注意:agents 表已删除 status 列(002),DB 不再持久化在线状态。
// 实际状态走 Redis presence key(hub.Register SET + 心跳 RefreshTTL 续期,60s TTL),
// agent_handler List/Get 时用 presence.IsOnline 实时算后赋值给 model.Agent.Status。
type AgentStatus string

const (
	AgentStatusOnline  AgentStatus = "online"
	AgentStatusOffline AgentStatus = "offline"
)

// AgentType 用类型常量集中定义 agent 类型标签。
// 底层 string,DB agents.type 列 VARCHAR(32),默认空串。
//
// 两大类(铺垫):
//   - 对话型(Hermes 类): 只看最终文本回复,单会话。
//     新 agent 走配对带 type=hermes;存量空串兼容归此类。
//   - 开发型(OpenCode 类): 多 session + 工具链 + 富 UI 卡片。
//     当前 opencode,Claude Code / Codex 等同类后期接入不再改源码判断。
type AgentType string

const (
	AgentTypeDefault  AgentType = ""         // 普通 agent(默认,存量空串,新版配对已不产生)
	AgentTypeHermes   AgentType = "hermes"   // hermes 对话型 agent(主流 IM 平台适配)
	AgentTypeOpencode AgentType = "opencode" // OpenCode 多 session 开发型 agent
)

type Agent struct {
	ID        string  `json:"id" db:"id"`
	OwnerID   string  `json:"owner_id" db:"owner_id"`
	Name      string  `json:"name" db:"name"`
	AvatarURL string  `json:"avatar_url" db:"avatar_url"`
	Bio       *string `json:"bio" db:"bio"`
	SecretKey string  `json:"-" db:"secret_key"`
	// Status 不映射 DB 列(002 已删),由 handler 用 presence 实时算后赋值。
	Status AgentStatus `json:"status" db:"-"`
	// Type 是 agent 类型标签(hermes 对话型 / opencode、dsh 开发型 / 空串 legacy 对话型)。
	Type AgentType `json:"type,omitempty" db:"type"`
	// MultiSession 是否多 session 拓扑(handler 按 type 查注册表填充,非 DB 列)。
	// APP 一级列表点击路由依据:true → 二级 sessions 页;false/nil → 直进聊天窗。
	// nil 兼容旧 server(字段缺失,APP 侧 fallback type=='opencode')。
	MultiSession *bool     `json:"multi_session,omitempty" db:"-"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
}

// AgentSummary 是 Agent 的展示型子集，用于 IM 列表等只需要展示信息的场景，
// 不暴露 secret_key / owner_id 等敏感或与展示无关的字段。
type AgentSummary struct {
	ID        string  `json:"id" db:"id"`
	Name      string  `json:"name" db:"name"`
	AvatarURL string  `json:"avatar_url" db:"avatar_url"`
	Bio       *string `json:"bio" db:"bio"`
	// Type 从 Agent 透传(应用层填充,非 DB 列)。供 IM 列表渲染/路由。
	Type AgentType `json:"type,omitempty" db:"-"`
	// MultiSession 从 Agent 透传(应用层按注册表填充,非 DB 列)。
	MultiSession *bool     `json:"multi_session,omitempty" db:"-"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
	// Status 在线状态(应用层用 hub.GetClient 实时算后赋值,非 DB 列)。
	Status AgentStatus `json:"status" db:"-"`
}

// ModelInfo 描述一个可选模型，由 plugin 上报、APP 用于模型选择器。
// 无 status 字段 — opencode 的 connected 状态目前不可信会断联。
type ModelInfo struct {
	ProviderID   string `json:"provider_id"`
	ProviderName string `json:"provider_name"`
	ModelID      string `json:"model_id"`
	ModelName    string `json:"model_name"`
}

// SlashCommandInfo 描述一条 OC 命令，由 plugin 上报、APP 用于斜杠命令列表。
// Source 区分 OC 命令(command)与 skill(skill)，APP 据此分组渲染。
type SlashCommandInfo struct {
	Name        string `json:"name"`
	Template    string `json:"template"`
	Description string `json:"description,omitempty"`
	Source      string `json:"source"`
}

type RpcMethod struct {
	Name          string `json:"name"`
	TimeoutHintMs int    `json:"timeout_hint_ms"`
}

// AgentModeInfo 描述一个会话模式，由 plugin 上报、APP 用于模式色条/选择器。
// Style 是受控渲染档位(default/plan/warn),APP 不理解 mode 的业务语义——
// 视觉差异由上报数据自描述(docs/research/proposal-agent-modes.md §3.5)。
type AgentModeInfo struct {
	ID    string `json:"id"`
	Label string `json:"label"`
	Style string `json:"style"`
}

// AgentPresetInfo 描述一个会话预设(能力组合)，由 plugin 上报。
// Trust 区分 system(部署方内置)/user(用户自创,权限等同 shell 访问),
// APP 据此显示来源标识。字段核对自 dsh agent-presets 真实 schema
// (dsh name → 本协议 label;order 排序位可选;broken 由桥接层过滤)。
type AgentPresetInfo struct {
	ID          string `json:"id"`
	Label       string `json:"label"`
	Description string `json:"description,omitempty"`
	Trust       string `json:"trust"`
	Order       int    `json:"order,omitempty"`
}

File: packages/wanling_core/lib/models/agent.dart
/// Agent 类型分类(集中 type 字面量 + 维度判断)。
///
/// 两大类(铺垫):
///   - 对话型(Hermes 类): 只看最终文本回复,单会话。
///   - 开发型(OpenCode 类): 多 session + 工具链 + 富 UI 卡片。
///     Claude Code / Codex 等同类后期接入只改 [supportsMultiSession]。
///
/// 存量 agent 的 type 可能为空串(老配对未上报 type),按对话型兼容。
abstract final class AgentCategory {
  static const hermes = 'hermes';
  static const opencode = 'opencode';

  /// 是否多 session(决定一级列表点击路由:走二级列表 vs 直进聊天窗)。
  /// 当前仅 opencode 为 true;Claude Code 等同类加入时在此扩展。
  static bool supportsMultiSession(String type) => type == opencode;
}

/// Agent 在线状态。后端 status 字段是 "online" / "offline" 字符串。
/// 用 enum 集中定义、IDE 友好、避免拼写错误。
enum AgentStatus {
  online,
  offline,
  unknown;
}

extension AgentStatusX on AgentStatus {
  String get value => switch (this) {
        AgentStatus.online => 'online',
        AgentStatus.offline => 'offline',
        AgentStatus.unknown => 'unknown',
      };

  static AgentStatus fromString(String? raw) {
    switch (raw) {
      case 'online':
        return AgentStatus.online;
      case 'offline':
        return AgentStatus.offline;
      default:
        return AgentStatus.unknown;
    }
  }
}

class Agent {
  final String id;
  final String name;
  final String? avatarUrl;
  final String? bio;
  final AgentStatus status;
  final String? secretKey;

  /// Agent 类型标签(参见 [AgentCategory])。
  /// 空串=legacy 对话型;新配对走 hermes/opencode 上报。
  final String type;

  /// 是否多 session 拓扑(server 按 type 注册表注入)。
  /// 决定一级列表点击路由:二级 sessions 页 vs 直进聊天窗。
  /// null=老 server 未下发(本地 fallback type=='opencode')。
  final bool? multiSession;

  Agent({
    required this.id,
    required this.name,
    this.avatarUrl,
    this.bio,
    required this.status,
    this.secretKey,
    this.type = '',
    this.multiSession,
  });

  factory Agent.fromJson(Map<String, dynamic> json) => Agent(
        id: json['id'],
        name: json['name'],
        avatarUrl: json['avatar_url'],
        bio: json['bio'],
        status: AgentStatusX.fromString(json['status']),
        secretKey: json['secret_key'],
        type: json['type'] as String? ?? '',
        multiSession: json['multi_session'] as bool?,
      );

  /// 多 session 路由判定:优先 server 注册表注入值,
  /// null(老 server 缺字段)时 fallback 旧口径 type=='opencode'。
  bool get isMultiSession => multiSession ?? type == AgentCategory.opencode;

  /// copyWith 用 clearBio bool 区分 bio "不动"和"清空"。
  Agent copyWith({
    String? name,
    String? avatarUrl,
    String? bio,
    bool clearBio = false,
    AgentStatus? status,
    String? type,
  }) =>
      Agent(
        id: id,
        name: name ?? this.name,
        avatarUrl: avatarUrl ?? this.avatarUrl,
        bio: clearBio ? null : (bio ?? this.bio),
        status: status ?? this.status,
        secretKey: secretKey,
        type: type ?? this.type,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'avatar_url': avatarUrl,
        'bio': bio,
        'status': status.value,
        'type': type,
      };
}

/// Agent 摘要(IM 列表 / 会话详情用)。
/// 与 [Agent] 区别:不含 secretKey,只渲染所需的最小字段。
/// bio 仅在需要展示简介的场景(如会话二级页 Drawer 头部)填充,其他可保持 null。
/// N 方 participants 模型下,仅 dm_user_agent 场景会填,其他 type 为 null。
class AgentSummary {
  final String id;
  final String name;
  final String? avatarUrl;
  final AgentStatus status;

  /// Agent 类型标签(参见 [AgentCategory])。
  /// 空串=legacy 对话型;新配对走 hermes/opencode 上报。
  final String type;

  /// 是否多 session 拓扑(server 按 type 注册表注入)。null=老 server 缺字段。
  final bool? multiSession;

  /// Agent 简介(可选)。仅二级页面渲染时透传,API JSON 默认不带。
  final String? bio;

  AgentSummary({
    required this.id,
    required this.name,
    this.avatarUrl,
    required this.status,
    this.type = '',
    this.multiSession,
    this.bio,
  });

  factory AgentSummary.fromJson(Map<String, dynamic> json) => AgentSummary(
        id: json['id'] as String,
        name: json['name'] as String,
        avatarUrl: json['avatar_url'] as String?,
        status: AgentStatusX.fromString(json['status'] as String?),
        type: json['type'] as String? ?? '',
        multiSession: json['multi_session'] as bool?,
        bio: json['bio'] as String?,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'avatar_url': avatarUrl,
        'status': status.value,
        'type': type,
        if (multiSession != null) 'multi_session': multiSession,
        if (bio != null) 'bio': bio,
      };

  /// 多 session 路由判定(与 [Agent.isMultiSession] 同口径)。
  bool get isMultiSession => multiSession ?? type == AgentCategory.opencode;
}

/// 可选模型条目(对应 server model.ModelInfo,Task 3 GET /api/agents/:id/models)。
/// 4 字段:provider/model id+name,无 status(模型不区分在线/离线)。
/// fromJson 容错(server 缺字段时回退 '',避免老 server 崩 client)。
class AgentModel {
  final String providerId;
  final String providerName;
  final String modelId;
  final String modelName;

  const AgentModel({
    required this.providerId,
    required this.providerName,
    required this.modelId,
    required this.modelName,
  });

  factory AgentModel.fromJson(Map<String, dynamic> json) {
    return AgentModel(
      providerId: json['provider_id'] as String? ?? '',
      providerName: json['provider_name'] as String? ?? '',
      modelId: json['model_id'] as String? ?? '',
      modelName: json['model_name'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'provider_id': providerId,
        'provider_name': providerName,
        'model_id': modelId,
        'model_name': modelName,
      };
}
